src/hepattn/flex/local_ca.py

```diff
@@ -0,0 +1,65 @@
+import torch
+from torch.nn.attention.flex_attention import BlockMask, _mask_mod_signature, create_block_mask
+
+
+@torch.compile(dynamic=True)
+def sliding_window_mask_strided(window_size: int, stride: float, q_len: int, kv_len: int, dev) -> _mask_mod_signature:
+    if window_size % 2 != 0:
+        raise ValueError("Window size must be even for strided sliding window")
+
+    stride_val = torch.tensor(stride, device=dev)
+
+    def mask_mod(b, h, q_idx, kv_idx):  # noqa: ARG001
+        # b = 0, h = 0 here
+        q_center = torch.round(q_idx * stride_val)
+        return (kv_idx - q_center).abs() <= window_size // 2
+
+    return create_block_mask(mask_mod, B=None, H=None, Q_LEN=q_len, KV_LEN=kv_len, device=dev)
+
+
+@torch.compile(dynamic=True)
+def sliding_window_mask_strided_wrapped(window_size: int, stride: float, q_len: int, kv_len: int, dev) -> _mask_mod_signature:
+    if window_size % 2 != 0:
+        raise ValueError("Window size must be even for strided sliding window")
+
+    stride_val = torch.tensor(stride, device=dev)
+    kv_len_t = torch.as_tensor(kv_len, device=dev).reshape(())
+
+    def mask_mod(b, h, q_idx, kv_idx):  # noqa: ARG001
+        # b = 0, h = 0 here
+        q_center = torch.round(q_idx * stride_val)
+
+        diagonal = (kv_idx - q_center).abs() <= window_size // 2
+        wrap_left = (kv_idx - q_center + kv_len_t).abs() <= window_size // 2
+        wrap_right = (kv_idx - q_center - kv_len_t).abs() <= window_size // 2
+
+        return diagonal | wrap_left | wrap_right
+
+    return create_block_mask(mask_mod, B=None, H=None, Q_LEN=q_len, KV_LEN=kv_len, device=dev)
+
+
+def transpose_blockmask(bm: BlockMask, *, q_tokens: int, kv_tokens: int, device=None) -> BlockMask:
+    """Exact transpose of a FlexAttention BlockMask by transposing the mask function.
+
+    Args:
+        bm: forward BlockMask (built with Q_LEN=q_tokens, KV_LEN=kv_tokens)
+        q_tokens: original forward Q token length
+        kv_tokens: original forward KV token length
+        device: torch.device to build the transposed mask on (defaults to cpu if None)
+    """
+    orig_mod = bm.mask_mod
+    dev = device if device is not None else getattr(bm, "device", "cpu")
```

Comment:
does the block mask not always have a device? in which case we probably don't need a dedicated `device` argument